mitiq/mitiq_qiskit/qiskit_utils.py

```diff
@@ -87,3 +89,182 @@ def run_with_noise(
     counts = results.get_counts()
     expval = counts["0"] / shots
     return expval
+
+
+def execute(circ: QuantumCircuit, obs: np.ndarray) -> float:
+    """Simulates noiseless wavefunction evolution and returns the
+    expectation value of some observable.
+
+    Args:
+        circ: The input Qiskit circuit.
+        obs: The observable to measure as a NumPy array.
+
+    Returns:
+        The expectation value of obs as a float.
+    """
+    result = qiskit.execute(circ, WVF_SIMULATOR).result()
+    final_wvf = result.get_statevector()
+    return np.real(final_wvf.conj().T @ obs @ final_wvf)
+
+
+def execute_with_shots(
+    circ: QuantumCircuit, obs: np.ndarray, shots: int
+) -> float:
+    """Simulates the evolution of the circuit and returns
+    the expectation value of the observable.
+
+    Args:
+        circ: The input Qiskit circuit.
+        obs: The observable to measure as a NumPy array.
+        shots: The number of measurements.
+
+    Returns:
+        The expectation value of obs as a float.
+
+    """
+    if len(circ.clbits) > 0:
+        raise ValueError(
+            "This executor only works on programs with no classical bits."
+        )
+
+    circ = copy.deepcopy(circ)
+    # we need to modify the circuit to measure obs in its eigenbasis
+    # we do this by appending a unitary operation
+    # obtains a U s.t. obs = U diag(eigvals) U^dag
+    eigvals, U = np.linalg.eigh(obs)
+    circ.unitary(np.linalg.inv(U), qubits=range(circ.num_qubits))
+
+    circ.measure_all()
+
+    # execution of the experiment
+    job = qiskit.execute(
+        circ,
+        backend=QASM_SIMULATOR,
+        # we want all gates to be actually applied,
+        # so we skip any circuit optimization
+        optimization_level=0,
+        shots=shots,
+    )
+    results = job.result()
+    counts = results.get_counts()
+    expectation = 0
+    # classical bits are included in bitstrings with a space
+    # this is what breaks if you have them
```

Comment:
I need to test this locally, I'm not sure this comment applies anymore. Even if it does, it would be better to remove the measurements from the circuit and run the rest of the function if possible.

Reply:
The bitstrings still return with a space, but easy enough to just do `expectation += eigvals[int(bitstring[0:circ.num_qubits], 2)] * count / shots`?

Reply:
If this approach works I would much prefer it to throwing an error whenever there are bits in the circuit!